### PR TITLE
Hivelord and drone's new ability: Recycler Redux

### PR DESCRIPTION
**Pull request by novaepee**
## About The Pull Request

#12309

## Why It's Good For The Game

"Dying as xenomorph always gave an fair more opressive feeling than dying as marine. Not only you lost time to age depending on your caste but you also gave a significant amount of points. The new skill recycle aims to give xenos the possibility to at least fight for the bodies of fallen xenos and try to recover a bit from the lost. Also this skill is a hivelord exclusive making the caste more wanted and interesting."

Bring back salvage bodymass, essentially. Lefunny didn't comply with Lumi's demand of reducing larva regen and sound choice, so PR was closed due to stale. We need xenomorphs content.

## Changelog

:cl: novaepee, Lefunny, Xander3359, OrcaCora, MacBlaze1
add: Hivelord and drone have a new ability, recycle
balance: Hivelord and drone can gib dead xenomorphs and minions to disrupt marine's req logistics
imageadd: recycle ability sprite
/:cl:

### CODE REVIEW

**Comment by EvilMelonTheSecond** [comment]
> We need xenomorphs content.

So buffs are content? Are you just saying random buzzwords to get this old mechanic merged? From what I remember salvage bodymass was REMOVED because the justification you said did not justify keeping it in the game. I do not mind the balance side of things but trying to sneak a balance PR as a "feature" or "content" PR bothers me.

File: code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm [hunk 1]
```diff
@@ -3,6 +3,78 @@
...
+	use_state_flags = XACT_USE_STAGGERED //can't use while staggered, defender fortified or crest down
+	keybinding_signals = list(
+		KEYBINDING_NORMAL = COMSIG_XENOABILITY_RECYCLE,
+	)
+	plasma_cost = 150
+	gamemode_flags = ABILITY_DISTRESS
+
+/datum/action/xeno_action/activable/recycle/can_use_ability(atom/target)
+	. = ..()
+	var/mob/living/carbon/xenomorph/X = owner
```
Comment (Lumipharon) [inline code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:22]: proper var names

File: code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm [hunk 2]
```diff
@@ -3,6 +3,78 @@
...
+		return FALSE
+	if(victim.stat != DEAD)
+		X.balloon_alert(X, "Sister isn't dead")
+		return FALSE
+
+	X.face_atom(victim) //Face towards the target so we don't look silly
+	X.visible_message(span_warning("\The [X] begins opening its mouth and extending a second jaw towards \the [victim]."), \
+	span_danger("We slowly feast upon \the [victim]'s carcass!"), null, 20)
+	var/channel = SSsounds.random_available_channel()
+	playsound(X, 'sound/vore/escape.ogg', 40, channel = channel)
```
Comment (Lumipharon) [inline code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:46]: pls no vore.

File: code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm [hunk 3]
```diff
@@ -3,6 +3,78 @@
...
+	if(!do_after(X, 7 SECONDS, FALSE, victim, BUSY_ICON_DANGER, extra_checks = CALLBACK(X, TYPE_PROC_REF(/mob, break_do_after_checks), list("health" = X.health))))
+		X.visible_message(span_xenowarning("\The [X] retracts its inner jaw."), \
+		span_danger("We retract our inner jaw."), null, 20)
+		X.stop_sound_channel(channel)
+		return FALSE
+	succeed_activate() //dew it
```
Comment (Lumipharon) [inline code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:52]: this should be in use_ability, not can_use

File: code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm [hunk 4]
```diff
@@ -3,6 +3,78 @@
...
+
+/datum/action/xeno_action/activable/recycle/use_ability(atom/target)
+	. = ..()
```
Comment (Lumipharon) [inline code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:55]: Doesn't need a call to parent

File: code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm [hunk 5]
```diff
@@ -3,6 +3,78 @@
...
+	if(victim.stat != DEAD)
+		X.balloon_alert(X, "Sister isn't dead")
+		return FALSE
+
+	X.face_atom(victim) //Face towards the target so we don't look silly
+	X.visible_message(span_warning("\The [X] begins opening its mouth and extending a second jaw towards \the [victim]."), \
+	span_danger("We slowly feast upon \the [victim]'s carcass!"), null, 20)
+	var/channel = SSsounds.random_available_channel()
+	playsound(X, 'sound/vore/escape.ogg', 40, channel = channel)
+	if(!do_after(X, 7 SECONDS, FALSE, victim, BUSY_ICON_DANGER, extra_checks = CALLBACK(X, TYPE_PROC_REF(/mob, break_do_after_checks), list("health" = X.health))))
```
Comment (Lumipharon) [inline code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:47]: Pretty sure that's not the right extra_check to use.

**Review by Lumipharon — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm [hunk 6]
```diff
@@ -3,6 +3,77 @@
...
+		if(XENO_TIER_THREE)
+			xeno_job.add_job_points(3)
+			recycled_xeno.gib()
+		if(XENO_TIER_FOUR)
+			xeno_job.add_job_points(4)
+			recycled_xeno.gib()
+
+	var/channel = SSsounds.random_available_channel()
+	playsound(hivelord, 'sound/effects/alien_recycler.ogg', 40, channel = channel)
+	hivelord.stop_sound_channel(channel)
```
Comment (ivanmixo) [inline code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:72]:
```suggestion
	playsound(hivelord, 'sound/effects/alien_recycler.ogg', 40)
```
What's the point of starting the sound only to stop it immediately?

**Comment by Pariah919** [comment]
This isn't even xenomorph content, its just powercreep. Being able to gib bodies is le big cringe considering its the main form of point gen for most marines you can already drag corpses, giving it to hivelord is even more cringe *as you already have so much utility in one caste.*

**Comment by RikuTheKiller** [comment]
> This isn't even xenomorph content, its just powercreep. Being able to gib bodies is le big cringe considering its the main form of point gen for most marines you can already drag corpses, giving it to hivelord is even more cringe _as you already have so much utility in one caste._

Just recover the corpses before a hivelord notices them. It's really not that hard considering most hives only have 1 or 2 hivelords and also hivelords spend a lot of their time building and backlining.

File: code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm [hunk 7]
```diff
@@ -3,6 +3,75 @@
...
+
+/datum/action/xeno_action/activable/recycle/can_use_ability(atom/target)
+	. = ..()
+	var/mob/living/carbon/xenomorph/hivelord = owner
+	var/mob/living/carbon/xenomorph/victim = target
+	if(!.)
+		return FALSE
+	if(hivelord.do_actions)
+		hivelord.balloon_alert(hivelord, "Busy")
+		return FALSE
```
Comment (Lumipharon) [inline code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:28]: This should already be checked by parent, unless you explicitly give it the override.

**Review by Lumipharon — CHANGES_REQUESTED** [review]

**Review by Lumipharon — APPROVED** [review]

File: code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm [hunk 8]
```diff
@@ -3,6 +3,72 @@
...
+
+/datum/action/xeno_action/activable/recycle/use_ability(atom/target)
+	var/mob/living/carbon/xenomorph/recycled_xeno = target
+	var/datum/job/xeno_job = SSjob.GetJobType(/datum/job/xenomorph)
+	var/mob/living/carbon/xenomorph/hivelord = owner
+	hivelord.face_atom(recycled_xeno) //Face towards the target so we don't look silly
+	hivelord.visible_message(span_warning("\The [hivelord] starts breaking apart \the [recycled_xeno]'s carcass."), \
+	span_danger("We slowly deconstruct upon \the [recycled_xeno]'s carcass!"), null, 20)
+	if(!do_after(owner, 7 SECONDS, FALSE, recycled_xeno, BUSY_ICON_GENERIC, extra_checks = CALLBACK(src, PROC_REF(can_use_ability))))
```
Comment (ivanmixo) [inline code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:46]:
```suggestion
	if(!do_after(owner, 7 SECONDS, FALSE, recycled_xeno, BUSY_ICON_GENERIC, extra_checks = CALLBACK(src, PROC_REF(can_use_ability), target, TRUE, XACT_USE_BUSY)))
```

File: code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm [hunk 9]
```diff
@@ -3,6 +3,72 @@
...
+
+/datum/action/xeno_action/activable/recycle/can_use_ability(atom/target)
+	. = ..()
+	var/mob/living/carbon/xenomorph/hivelord = owner
+	var/mob/living/carbon/xenomorph/victim = target
+	if(!.)
+		return FALSE
+	if(!hivelord.Adjacent(victim))
+		hivelord.balloon_alert(hivelord, "Too far")
+		return FALSE
```
Comment (ivanmixo) [inline code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:28]:
```suggestion
		if(!silent)
				hivelord.balloon_alert(hivelord, "Too far")
		return FALSE
```

File: code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm [hunk 10]
```diff
@@ -3,6 +3,72 @@
...
+	if(hivelord.on_fire)
+		hivelord.balloon_alert(hivelord, "Cannot while burning")
+		return FALSE
```
Comment (ivanmixo) [inline code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:31]:
```suggestion
		if(!silent)
				hivelord.balloon_alert(hivelord, "Cannot while burning")
		return FALSE
```

File: code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm [hunk 11]
```diff
@@ -3,6 +3,72 @@
...
+	if(!iscarbon(target)||!isxeno(target))
+		hivelord.balloon_alert(hivelord, "Cannot recycle")
+		return FALSE
+	if(victim.stat != DEAD)
+		hivelord.balloon_alert(hivelord, "Sister isn't dead")
```
Comment (ivanmixo) [inline code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:36]:
```suggestion
	if(victim.stat != DEAD)
		if(!silent)
			hivelord.balloon_alert(hivelord, "Sister isn't dead")
```

File: code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm [hunk 12]
```diff
@@ -3,6 +3,72 @@
...
+	if(!.)
+		return FALSE
+	if(!hivelord.Adjacent(victim))
+		hivelord.balloon_alert(hivelord, "Too far")
+		return FALSE
+	if(hivelord.on_fire)
+		hivelord.balloon_alert(hivelord, "Cannot while burning")
+		return FALSE
+	if(!iscarbon(target)||!isxeno(target))
+		hivelord.balloon_alert(hivelord, "Cannot recycle")
```
Comment (ivanmixo) [inline code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:33]:
```suggestion
	if(!iscarbon(target)||!isxeno(target))
		if(!silent)
			hivelord.balloon_alert(hivelord, "Cannot recycle")
```

File: code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm [hunk 13]
```diff
@@ -3,6 +3,72 @@
...
+	action_icon_state = "recycle"
+	desc = "We deconstruct the body of a fellow fallen xenomorph to avoid marines from harvesting our sisters in arms. Gain back a few larva points."
+	use_state_flags = XACT_USE_STAGGERED //can't use while staggered, defender fortified or crest down
+	keybinding_signals = list(
+		KEYBINDING_NORMAL = COMSIG_XENOABILITY_RECYCLE,
+	)
+	plasma_cost = 150
+	gamemode_flags = ABILITY_DISTRESS
+
+/datum/action/xeno_action/activable/recycle/can_use_ability(atom/target)
```
Comment (ivanmixo) [inline code/modules/mob/living/carbon/xenomorph/castes/hivelord/abilities_hivelord.dm:20]:
```suggestion
/datum/action/xeno_action/activable/recycle/can_use_ability(atom/target, silent = FALSE, override_flags)
```

**Review by ivanmixo — COMMENTED** [review]
Should fix it being broken

**Review by Lumipharon — APPROVED** [review]

**Comment by Naaanii** [comment]
Kill larva point refund or the gibbing.

**Comment by Lumipharon** [comment]
I'm kinda eh about this.

As it is, the ability is very strong overall, and represents a significant increase in larva generation over time, on top of reducing marine req point production significantly.

I don't think such a larve balance change is really needed, and if it wasn't it really shouldn't be focused on a single caste like this, as it really does make them a 'mandatory' pick.

**Comment by maikilangiolo** [comment]
Deny points *and* get larva points back? On the hivelord too, which already is very strong. Kinda cringe

**Comment by novaepee** [comment]
> Deny points _and_ get larva points back? On the hivelord too, which already is very strong. Kinda cringe

The larva points got removal if you read the PR.

**Review by Lumipharon — APPROVED** [review]

**Pull request merged**